# # Prokaryotic functional genome annotation

### So far have obtained indiviual metagenome assembled genomes (MAGs) from the rummen metagenome. Let's review our workflow to see what is the next step: 

![workflow](https://github.com/avera1988/NMBU-Bio-326/blob/main/images/wrokflowmetagenome.png) 

After binning with [MetaBat2](https://bitbucket.org/berkeleylab/metabat/src/master/) we obtained 9 bins. We used [CheckM](https://ecogenomics.github.io/CheckM/) to asses the quality of these MAGs.

Metabat2 produce a set of bins as the following example:

```bash
ls -1|grep .fa|sort -V
bin.1.fa
bin.2.fa
bin.3.fa
bin.4.fa
bin.5.fa
bin.6.fa
bin.7.fa
bin.8.fa
bin.9.fa
bin.10.fa
bin.11.fa
bin.12.fa
bin.13.fa
bin.14.fa
bin.15.fa
bin.16.fa
bin.17.fa
bin.18.fa
bin.19.fa
bin.20.fa
bin.21.fa
bin.22.fa
bin.23.fa
bin.24.fa
bin.25.fa
bin.26.fa
```
We then can use the report from the ```assemblycomparator2``` to check the quality report by CheckM. We can use either the report table or directly the checkM results 

Let's go to the ```cd results_ac2/``` folder and enter the checkM results

```console
$ ls results_ac2/checkm2/
checkm2.log  diamond_output  protein_files  quality_report.tsv
$ more results_ac2/checkm2/quality_report.tsv
Name    Completeness    Contamination   Completeness_Model_Used Translation_Table_Used  Coding_Density  Contig_N50      Aver
age_Gene_Length Genome_Size     GC_Content      Total_Coding_Sequences  Additional_Notes
bin.1   65.29   12.04   Gradient Boost (General Model)  11      0.824   61619   223.82011834319528      1368180 0.54    1690
        None
bin.10  88.2    20.77   Neural Network (Specific Model) 11      0.887   125057  296.9914163090129       2335282 0.33    2330
        None
bin.11  93.91   63.86   Gradient Boost (General Model)  11      0.848   27951   224.5304107844178       7434991 0.5     9421
        Low confidence prediction - substantial (34%) disagreement between completeness prediction models
bin.12  50.41   31.62   Neural Network (Specific Model) 11      0.735   29328   192.14297589359933      2814547 0.31    3609
        None
bin.13  67.87   3.91    Neural Network (Specific Model) 11      0.9     1187215 276.24345364472754      1298639 0.37    1413
        None
```

Now can start filtering the MAGs. A good criterion is to use the quality and contamination of the MAGs to sort them into *High*, *Medium* and *Low* quality MAGs. 
We can use the following table from [Bowers et al.,](https://www.nature.com/articles/nbt.3893) to classify the MAGs:
![tablemags](https://github.com/avera1988/NMBU-Bio-326/blob/main/images/mags.jpg)

We have all this information from the table generated by checkm, and we can use it to extract these quality score parameters. We can easily pick manually those "good quality" MAGs and sorted them for quality and contamination scores. Let's say >= 70 % completeness and =< 10 % contamination by looking into this table and applying some conditional using awk:

```bash
cat results_ac2/checkm2/quality_report.tsv | awk '{if($2 > 70 && $3 < 10) print $1"\t"$2"\t"$3}'
rumen.3 94.14   7.56
rumen.6 72.12   1.12
```
**We can see here that only 2 bins meet the condition of > 70 % completeness and < 10 % contamination.**

Then let's look for the taxonomy of those particularly bins using the gtdbtk results:

```bash
cat results_ac2/checkm2/quality_report.tsv | awk '{if($2 > 70 && $3 < 10) print $1}'|sed 's/.f
a//g'|fgrep -f - results_ac2/gtdbtk/gtdbtk.summary.tsv |cut -f 1,2
rumen.3 d__Archaea;p__Methanobacteriota;c__Methanobacteria;o__Methanobacteriales;f__Methanobacteriaceae;g__Methanobrevibacter_A;s__Methanobrevibacter_A sp900313645
rumen.6 d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Prevotella;s__Prevotella sp900316295
```

**Our bins were classified as an Archea *Metanobrevibacter* and a bacteria *Prevotella*.**

We can then predict genes and annotate this by comparing with public databases. For doing this we will use DRAM...

## DRAM: Distilled and Refined Annotation of Metabolism

"[DRAM](https://github.com/WrightonLabCSU/DRAM) (Distilled and Refined Annotation of Metabolism) is a tool for annotating metagenomic assembled genomes and VirSorter identified viral contigs. DRAM annotates MAGs and viral contigs using KEGG (if provided by the user), UniRef90, PFAM, dbCAN, RefSeq viral, VOGDB and the MEROPS peptidase database as well as custom user databases..."
 ![dramaanot](https://github.com/avera1988/NMBU-Bio-326/blob/main/images/DRAM.jpg)
