# Assembly a bacterial genome using long reads generated by the Oxford Nanopore MiniON and PromethiON platforms![ONP](https://github.com/TheMEMOLab/Bio326-NMBU/blob/main/images/ONP.png).

 **After preparing the sequencing library and load it on a MiniON and PromethiON flowcells, [Guppy](https://nanoporetech.com/nanopore-sequencing-data-analysis) was used for basecalling. 
 Fastq files were obtained of this sequencing experiment. The following protocol describes the steps for assembly those long-ONP reads into a bacterial genome**
 
 ## Obtaining the fastq files
 
 * Loggin to Orion cluster

```console
ssh bio326-21-0@login.orion.nmbu.no
```

* Ceate a directory in your $SCRATCH folder named ```GenomeAssembly2022```. Then copy the MiniON and PromethiON reads from ```/mnt/SCRATCH/bio326-21/GenomeAssembly/BIO326-2022```

```console
[bio326-21-0@login bio326-21-0]$ mkdir GenomeAssembly2022
[bio326-21-0@login bio326-21-0]$ cd GenomeAssembly2022
[bio326-21-0@login GenomeAssembly2022]$ rsync -aP /mnt/SCRATCH/bio326-21/GenomeAssembly/BIO326-2022/MiniONReads .
[bio326-21-0@login GenomeAssembly2022]$ rsync -aP /mnt/SCRATCH/bio326-21/GenomeAssembly/BIO326-2022/PromethiONReads/ .
[bio326-21-0@login GenomeAssembly2022]$ ls
MiniONReads  PromethiONReads
```

* Let's take a look into the folders:

```console
[bio326-21-0@login GenomeAssembly2022]$ tree *
MiniONReads
└── fastqFiles
    ├── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_0_0.fastq
    ├── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_1_0.fastq
    ├── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_2_0.fastq
    ├── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_3_0.fastq
    ├── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_4_0.fastq
    └── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_5_0.fastq
PromethiONReads
└── fastqFiles
    ├── PAG41242_pass_barcode26_24f50ffb_0.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_10.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_11.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_12.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_13.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_14.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_15.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_16.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_17.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_18.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_19.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_1.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_20.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_21.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_22.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_23.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_24.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_25.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_26.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_27.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_28.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_29.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_2.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_30.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_31.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_32.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_33.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_34.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_35.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_36.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_37.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_38.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_39.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_3.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_40.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_41.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_4.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_5.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_6.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_7.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_8.fastq
    └── PAG41242_pass_barcode26_24f50ffb_9.fastq

2 directories, 48 files
```

* As you notice there are many fastq files in the directories. **It is often useful to concatenate (merge) all the different fastq files into one big file for downstream analyses.** The following ```for loop``` will be useful:


```console
[bio326-21-0@login GenomeAssembly2022]$ for i in *Reads; do Prefix=$(basename $i Reads); cat $i/fastqFiles/*fastq > $i/$Prefix.fastq;done
```
*Using ```for loops``` is a great tool to analyze multiple files/directories at once. In this [link](https://devhints.io/bash) you will find a cheatsheet with many tricks on bash programming...*

Let's take a look on the results of our ```for loop```:

```console
[bio326-21-0@login GenomeAssembly2022]$ tree *
MiniONReads
├── fastqFiles
│   ├── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_0_0.fastq
│   ├── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_1_0.fastq
│   ├── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_2_0.fastq
│   ├── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_3_0.fastq
│   ├── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_4_0.fastq
│   └── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_5_0.fastq
└── MiniON.fastq
PromethiONReads
├── fastqFiles
│   ├── PAG41242_pass_barcode26_24f50ffb_0.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_10.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_11.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_12.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_13.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_14.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_15.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_16.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_17.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_18.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_19.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_1.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_20.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_21.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_22.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_23.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_24.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_25.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_26.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_27.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_28.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_29.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_2.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_30.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_31.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_32.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_33.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_34.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_35.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_36.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_37.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_38.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_39.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_3.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_40.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_41.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_4.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_5.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_6.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_7.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_8.fastq
│   └── PAG41242_pass_barcode26_24f50ffb_9.fastq
└── PromethiON.fastq

2 directories, 50 files
[bio326-21-0@login GenomeAssembly2022]$ tree -du *
MiniONReads
└── [bio326-2]  fastqFiles
PromethiONReads
└── [bio326-2]  fastqFiles

2 directories
[bio326-21-0@login GenomeAssembly2022]$ tree *
MiniONReads
├── fastqFiles
│   ├── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_0_0.fastq
│   ├── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_1_0.fastq
│   ├── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_2_0.fastq
│   ├── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_3_0.fastq
│   ├── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_4_0.fastq
│   └── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_5_0.fastq
└── MiniON.fastq
PromethiONReads
├── fastqFiles
│   ├── PAG41242_pass_barcode26_24f50ffb_0.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_10.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_11.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_12.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_13.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_14.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_15.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_16.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_17.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_18.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_19.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_1.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_20.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_21.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_22.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_23.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_24.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_25.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_26.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_27.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_28.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_29.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_2.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_30.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_31.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_32.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_33.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_34.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_35.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_36.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_37.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_38.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_39.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_3.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_40.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_41.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_4.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_5.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_6.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_7.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_8.fastq
│   └── PAG41242_pass_barcode26_24f50ffb_9.fastq
└── PromethiON.fastq

2 directories, 50 files
[bio326-21-0@login GenomeAssembly2022]$ tree -sh *
MiniONReads
├── [4.0K]  fastqFiles
│   ├── [ 32M]  fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_0_0.fastq
│   ├── [ 31M]  fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_1_0.fastq
│   ├── [ 31M]  fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_2_0.fastq
│   ├── [ 34M]  fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_3_0.fastq
│   ├── [ 35M]  fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_4_0.fastq
│   └── [ 36M]  fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_5_0.fastq
└── [199M]  MiniON.fastq
PromethiONReads
├── [4.0K]  fastqFiles
│   ├── [ 43M]  PAG41242_pass_barcode26_24f50ffb_0.fastq
│   ├── [ 52M]  PAG41242_pass_barcode26_24f50ffb_10.fastq
│   ├── [ 52M]  PAG41242_pass_barcode26_24f50ffb_11.fastq
│   ├── [ 57M]  PAG41242_pass_barcode26_24f50ffb_12.fastq
│   ├── [ 53M]  PAG41242_pass_barcode26_24f50ffb_13.fastq
│   ├── [ 55M]  PAG41242_pass_barcode26_24f50ffb_14.fastq
│   ├── [ 56M]  PAG41242_pass_barcode26_24f50ffb_15.fastq
│   ├── [ 54M]  PAG41242_pass_barcode26_24f50ffb_16.fastq
│   ├── [ 54M]  PAG41242_pass_barcode26_24f50ffb_17.fastq
│   ├── [ 54M]  PAG41242_pass_barcode26_24f50ffb_18.fastq
│   ├── [ 54M]  PAG41242_pass_barcode26_24f50ffb_19.fastq
│   ├── [ 48M]  PAG41242_pass_barcode26_24f50ffb_1.fastq
│   ├── [ 53M]  PAG41242_pass_barcode26_24f50ffb_20.fastq
│   ├── [ 54M]  PAG41242_pass_barcode26_24f50ffb_21.fastq
│   ├── [ 54M]  PAG41242_pass_barcode26_24f50ffb_22.fastq
│   ├── [ 56M]  PAG41242_pass_barcode26_24f50ffb_23.fastq
│   ├── [ 54M]  PAG41242_pass_barcode26_24f50ffb_24.fastq
│   ├── [ 56M]  PAG41242_pass_barcode26_24f50ffb_25.fastq
│   ├── [ 59M]  PAG41242_pass_barcode26_24f50ffb_26.fastq
│   ├── [ 57M]  PAG41242_pass_barcode26_24f50ffb_27.fastq
│   ├── [ 56M]  PAG41242_pass_barcode26_24f50ffb_28.fastq
│   ├── [ 55M]  PAG41242_pass_barcode26_24f50ffb_29.fastq
│   ├── [ 50M]  PAG41242_pass_barcode26_24f50ffb_2.fastq
│   ├── [ 57M]  PAG41242_pass_barcode26_24f50ffb_30.fastq
│   ├── [ 56M]  PAG41242_pass_barcode26_24f50ffb_31.fastq
│   ├── [ 56M]  PAG41242_pass_barcode26_24f50ffb_32.fastq
│   ├── [ 57M]  PAG41242_pass_barcode26_24f50ffb_33.fastq
│   ├── [ 56M]  PAG41242_pass_barcode26_24f50ffb_34.fastq
│   ├── [ 53M]  PAG41242_pass_barcode26_24f50ffb_35.fastq
│   ├── [ 58M]  PAG41242_pass_barcode26_24f50ffb_36.fastq
│   ├── [ 59M]  PAG41242_pass_barcode26_24f50ffb_37.fastq
│   ├── [ 56M]  PAG41242_pass_barcode26_24f50ffb_38.fastq
│   ├── [ 55M]  PAG41242_pass_barcode26_24f50ffb_39.fastq
│   ├── [ 49M]  PAG41242_pass_barcode26_24f50ffb_3.fastq
│   ├── [ 59M]  PAG41242_pass_barcode26_24f50ffb_40.fastq
│   ├── [ 19M]  PAG41242_pass_barcode26_24f50ffb_41.fastq
│   ├── [ 49M]  PAG41242_pass_barcode26_24f50ffb_4.fastq
│   ├── [ 50M]  PAG41242_pass_barcode26_24f50ffb_5.fastq
│   ├── [ 48M]  PAG41242_pass_barcode26_24f50ffb_6.fastq
│   ├── [ 50M]  PAG41242_pass_barcode26_24f50ffb_7.fastq
│   ├── [ 52M]  PAG41242_pass_barcode26_24f50ffb_8.fastq
│   └── [ 52M]  PAG41242_pass_barcode26_24f50ffb_9.fastq
└── [2.2G]  PromethiON.fastq

2 directories, 50 files
```
Now we can delete the fastqFiles directory in each folder:

```console
[bio326-21-0@login GenomeAssembly2022]$ for i in *Reads; do rm -r $i/fastqFiles;done
[bio326-21-0@login GenomeAssembly2022]$ tree -sh
.
├── [  33]  MiniONReads
│   └── [199M]  MiniON.fastq
└── [  37]  PromethiONReads
    └── [2.2G]  PromethiON.fastq

2 directories, 2 files
```

## Quality control check of the fastq files using [NanoPlot](https://github.com/wdecoster/NanoPlot).

The first steep after obtaining the fastq files form the sequencer is to know how is the quality in means of No. of reads, size, and ![Phred quality score](https://github.com/TheMEMOLab/Bio326-NMBU/blob/main/images/fastqC.png):

* We will use the [NanoPlot](https://github.com/wdecoster/NanoPlot) and some Rscripts to produce some plots of the basic stats of our data. To use Nanoplot we need first to request an interactive job and start working there. As some times we could lost the connection with the Orion server, it is preferable to use [TMUX](https://github.com/tmux/tmux/wiki). This terminal multiplexer software will allow keeping our session active in Orion even if we suddenly lost the internet connection:

```console
[bio326-21-0@login GenomeAssembly2022]$ tmux new -s ONP
Welcome to the NMBU Orion compute cluster environment.

You are logged in to a machine that can be used to access your home directory,
edit your scripts, manage your files, and submit jobs to the cluster environment.
Do not run any jobs on this machine, as they might be automatically terminated.

IMPORTANT:
  - Orion introduction: https://orion.nmbu.no/
  - Orion can handle small-scale projects. Need more CPU hours? Please consider
    applying for national infrastructure resources: https://www.sigma2.no/
  - Please, PLEASE do compress your fastq, vcf and other non-compressed files
    using i.e. pigz.

For any Orion related enquiry: orion-support@nmbu.no
PS: We are on Teams: https://bit.ly/orion-teams

[bio326-21-0@login GenomeAssembly2022]$
```
* Now we need to ask the SLURM manager for a computer node to test the NANOPLOT and can start to work: 

```console 
[bio326-21-0@login GenomeAssembly2022]$ srun --cpus-per-task 4 --nodes 1 --mem=10G --time=02:00:00 --pty bash -i
srun: job 14301977 queued and waiting for resources
srun: job 14301977 has been allocated resources

Welcome to the NMBU Orion compute cluster environment.

You are logged in to a machine that can be used to access your home directory,
edit your scripts, manage your files, and submit jobs to the cluster environment.
Do not run any jobs on this machine, as they might be automatically terminated.

IMPORTANT:
  - Orion introduction: https://orion.nmbu.no/
  - Orion can handle small-scale projects. Need more CPU hours? Please consider
    applying for national infrastructure resources: https://www.sigma2.no/
  - Please, PLEASE do compress your fastq, vcf and other non-compressed files
    using i.e. pigz.

NEWS:
  - 2020-10-08: Orion has been re-built. We are still working out many details.
    Please email us if you miss anything, or notice any issues.

For any Orion related enquiry: orion-support@nmbu.no
PS: We are on Teams: https://bit.ly/orion-teams

[bio326-21-0@cn-12 GenomeAssembly2022]$
```
*Now that TMUX is active you can deatch the temrinal and try if your session is still active. To do this we need to press the keys ctrl+b d. This will bring us back to the "main" terminal...Use the command ```tmux a -t ONP``` to attach the ONP session again*




