# Assembly a bacterial genome using long reads generated by the Oxford Nanopore MiniON and PromethiON platforms![ONP](https://github.com/TheMEMOLab/Bio326-NMBU/blob/main/images/ONP.png).

 **After preparing the sequencing library and load it on a MiniON and PromethiON flowcells, [Guppy](https://nanoporetech.com/nanopore-sequencing-data-analysis) was used for basecalling. 
 Fastq files were obtained of this sequencing experiment. The following protocol describes the steps for assembly those long-ONP reads into a bacterial genome**
 
 ## Obtaining the fastq files
 
 * Loggin to Orion cluster

```console
ssh bio326-21-0@login.orion.nmbu.no
```

* Ceate a directory in your $SCRATCH folder named ```GenomeAssembly2022```. Then copy the MiniON and PromethiON reads from ```/mnt/SCRATCH/bio326-21/GenomeAssembly/BIO326-2022```

```console
[bio326-21-0@login bio326-21-0]$ mkdir GenomeAssembly2022
[bio326-21-0@login bio326-21-0]$ cd GenomeAssembly2022
[bio326-21-0@login GenomeAssembly2022]$ rsync -aP /mnt/SCRATCH/bio326-21/GenomeAssembly/BIO326-2022/MiniONReads .
[bio326-21-0@login GenomeAssembly2022]$ rsync -aP /mnt/SCRATCH/bio326-21/GenomeAssembly/BIO326-2022/PromethiONReads/ .
[bio326-21-0@login GenomeAssembly2022]$ ls
MiniONReads  PromethiONReads
```

* Let's take a look into the folders:

```console
[bio326-21-0@login GenomeAssembly2022]$ tree *
MiniONReads
└── fastqFiles
    ├── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_0_0.fastq
    ├── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_1_0.fastq
    ├── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_2_0.fastq
    ├── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_3_0.fastq
    ├── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_4_0.fastq
    └── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_5_0.fastq
PromethiONReads
└── fastqFiles
    ├── PAG41242_pass_barcode26_24f50ffb_0.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_10.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_11.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_12.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_13.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_14.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_15.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_16.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_17.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_18.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_19.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_1.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_20.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_21.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_22.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_23.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_24.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_25.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_26.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_27.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_28.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_29.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_2.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_30.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_31.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_32.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_33.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_34.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_35.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_36.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_37.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_38.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_39.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_3.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_40.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_41.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_4.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_5.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_6.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_7.fastq
    ├── PAG41242_pass_barcode26_24f50ffb_8.fastq
    └── PAG41242_pass_barcode26_24f50ffb_9.fastq

2 directories, 48 files
```

* As you notice there are many fastq files in the directories. **It is often useful to concatenate (merge) all the different fastq files into one big file for downstream analyses.** The following ```for loop``` will be useful:


```console
[bio326-21-0@login GenomeAssembly2022]$ for i in *Reads; do Prefix=$(basename $i Reads); cat $i/fastqFiles/*fastq > $i/$Prefix.fastq;done
```
*Using ```for loops``` is a great tool to analyze multiple files/directories at once. In this [link](https://devhints.io/bash) you will find a cheatsheet with many tricks on bash programming...*

Let's take a look on the results of our ```for loop```:

```console
[bio326-21-0@login GenomeAssembly2022]$ tree *
MiniONReads
├── fastqFiles
│   ├── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_0_0.fastq
│   ├── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_1_0.fastq
│   ├── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_2_0.fastq
│   ├── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_3_0.fastq
│   ├── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_4_0.fastq
│   └── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_5_0.fastq
└── MiniON.fastq
PromethiONReads
├── fastqFiles
│   ├── PAG41242_pass_barcode26_24f50ffb_0.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_10.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_11.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_12.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_13.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_14.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_15.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_16.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_17.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_18.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_19.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_1.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_20.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_21.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_22.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_23.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_24.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_25.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_26.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_27.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_28.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_29.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_2.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_30.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_31.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_32.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_33.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_34.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_35.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_36.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_37.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_38.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_39.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_3.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_40.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_41.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_4.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_5.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_6.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_7.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_8.fastq
│   └── PAG41242_pass_barcode26_24f50ffb_9.fastq
└── PromethiON.fastq

2 directories, 50 files
[bio326-21-0@login GenomeAssembly2022]$ tree -du *
MiniONReads
└── [bio326-2]  fastqFiles
PromethiONReads
└── [bio326-2]  fastqFiles

2 directories
[bio326-21-0@login GenomeAssembly2022]$ tree *
MiniONReads
├── fastqFiles
│   ├── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_0_0.fastq
│   ├── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_1_0.fastq
│   ├── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_2_0.fastq
│   ├── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_3_0.fastq
│   ├── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_4_0.fastq
│   └── fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_5_0.fastq
└── MiniON.fastq
PromethiONReads
├── fastqFiles
│   ├── PAG41242_pass_barcode26_24f50ffb_0.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_10.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_11.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_12.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_13.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_14.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_15.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_16.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_17.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_18.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_19.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_1.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_20.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_21.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_22.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_23.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_24.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_25.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_26.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_27.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_28.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_29.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_2.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_30.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_31.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_32.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_33.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_34.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_35.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_36.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_37.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_38.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_39.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_3.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_40.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_41.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_4.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_5.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_6.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_7.fastq
│   ├── PAG41242_pass_barcode26_24f50ffb_8.fastq
│   └── PAG41242_pass_barcode26_24f50ffb_9.fastq
└── PromethiON.fastq

2 directories, 50 files
[bio326-21-0@login GenomeAssembly2022]$ tree -sh *
MiniONReads
├── [4.0K]  fastqFiles
│   ├── [ 32M]  fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_0_0.fastq
│   ├── [ 31M]  fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_1_0.fastq
│   ├── [ 31M]  fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_2_0.fastq
│   ├── [ 34M]  fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_3_0.fastq
│   ├── [ 35M]  fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_4_0.fastq
│   └── [ 36M]  fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_5_0.fastq
└── [199M]  MiniON.fastq
PromethiONReads
├── [4.0K]  fastqFiles
│   ├── [ 43M]  PAG41242_pass_barcode26_24f50ffb_0.fastq
│   ├── [ 52M]  PAG41242_pass_barcode26_24f50ffb_10.fastq
│   ├── [ 52M]  PAG41242_pass_barcode26_24f50ffb_11.fastq
│   ├── [ 57M]  PAG41242_pass_barcode26_24f50ffb_12.fastq
│   ├── [ 53M]  PAG41242_pass_barcode26_24f50ffb_13.fastq
│   ├── [ 55M]  PAG41242_pass_barcode26_24f50ffb_14.fastq
│   ├── [ 56M]  PAG41242_pass_barcode26_24f50ffb_15.fastq
│   ├── [ 54M]  PAG41242_pass_barcode26_24f50ffb_16.fastq
│   ├── [ 54M]  PAG41242_pass_barcode26_24f50ffb_17.fastq
│   ├── [ 54M]  PAG41242_pass_barcode26_24f50ffb_18.fastq
│   ├── [ 54M]  PAG41242_pass_barcode26_24f50ffb_19.fastq
│   ├── [ 48M]  PAG41242_pass_barcode26_24f50ffb_1.fastq
│   ├── [ 53M]  PAG41242_pass_barcode26_24f50ffb_20.fastq
│   ├── [ 54M]  PAG41242_pass_barcode26_24f50ffb_21.fastq
│   ├── [ 54M]  PAG41242_pass_barcode26_24f50ffb_22.fastq
│   ├── [ 56M]  PAG41242_pass_barcode26_24f50ffb_23.fastq
│   ├── [ 54M]  PAG41242_pass_barcode26_24f50ffb_24.fastq
│   ├── [ 56M]  PAG41242_pass_barcode26_24f50ffb_25.fastq
│   ├── [ 59M]  PAG41242_pass_barcode26_24f50ffb_26.fastq
│   ├── [ 57M]  PAG41242_pass_barcode26_24f50ffb_27.fastq
│   ├── [ 56M]  PAG41242_pass_barcode26_24f50ffb_28.fastq
│   ├── [ 55M]  PAG41242_pass_barcode26_24f50ffb_29.fastq
│   ├── [ 50M]  PAG41242_pass_barcode26_24f50ffb_2.fastq
│   ├── [ 57M]  PAG41242_pass_barcode26_24f50ffb_30.fastq
│   ├── [ 56M]  PAG41242_pass_barcode26_24f50ffb_31.fastq
│   ├── [ 56M]  PAG41242_pass_barcode26_24f50ffb_32.fastq
│   ├── [ 57M]  PAG41242_pass_barcode26_24f50ffb_33.fastq
│   ├── [ 56M]  PAG41242_pass_barcode26_24f50ffb_34.fastq
│   ├── [ 53M]  PAG41242_pass_barcode26_24f50ffb_35.fastq
│   ├── [ 58M]  PAG41242_pass_barcode26_24f50ffb_36.fastq
│   ├── [ 59M]  PAG41242_pass_barcode26_24f50ffb_37.fastq
│   ├── [ 56M]  PAG41242_pass_barcode26_24f50ffb_38.fastq
│   ├── [ 55M]  PAG41242_pass_barcode26_24f50ffb_39.fastq
│   ├── [ 49M]  PAG41242_pass_barcode26_24f50ffb_3.fastq
│   ├── [ 59M]  PAG41242_pass_barcode26_24f50ffb_40.fastq
│   ├── [ 19M]  PAG41242_pass_barcode26_24f50ffb_41.fastq
│   ├── [ 49M]  PAG41242_pass_barcode26_24f50ffb_4.fastq
│   ├── [ 50M]  PAG41242_pass_barcode26_24f50ffb_5.fastq
│   ├── [ 48M]  PAG41242_pass_barcode26_24f50ffb_6.fastq
│   ├── [ 50M]  PAG41242_pass_barcode26_24f50ffb_7.fastq
│   ├── [ 52M]  PAG41242_pass_barcode26_24f50ffb_8.fastq
│   └── [ 52M]  PAG41242_pass_barcode26_24f50ffb_9.fastq
└── [2.2G]  PromethiON.fastq

2 directories, 50 files
```
Now we can delete the fastqFiles directory in each folder:

```console
[bio326-21-0@login GenomeAssembly2022]$ for i in *Reads; do rm -r $i/fastqFiles;done
[bio326-21-0@login GenomeAssembly2022]$ tree -sh
.
├── [  33]  MiniONReads
│   └── [199M]  MiniON.fastq
└── [  37]  PromethiONReads
    └── [2.2G]  PromethiON.fastq

2 directories, 2 files
```

## Quality control check of the fastq files using [NanoPlot](https://github.com/wdecoster/NanoPlot).

The first steep after obtaining the fastq files form the sequencer is to know how is the quality in means of No. of reads, size, and ![Phred quality score](https://github.com/TheMEMOLab/Bio326-NMBU/blob/main/images/fastqC.png):

* Nanoplot and most of the software we will use in this course are installed in a condaenvironment. Let's activate the environment and test Nanoplot.

**NOTE: CONDA and all its features have to be loaded by a module named ```Anaconda3``` and then configure some of the environment variables to be able to properly use conda. For your convenience we have crated a bash script that does all this for you, just type the following to activate conda:**

```console
[bio326-21-0@login GenomeAssembly2022]$ source /mnt/SCRATCH/bio326-21/GenomeAssembly/condaenvironments/activate.conda.sh
Activating Anaconda module for bio326-21-0
conda is running. Please type conda activate to load the basic conda functions...
[bio326-21-0@login GenomeAssembly2022]$
```

Now that conda is loaded let's activate the ONPTools environment and try Nanoplot

```console
[bio326-21-0@login GenomeAssembly2022]$ conda activate  /mnt/SCRATCH/bio326-21/GenomeAssembly/condaenvironments/ONPTools/
(/mnt/SCRATCH/bio326-21/GenomeAssembly/condaenvironments/ONPTools) [bio326-21-0@login GenomeAssembly2022]$ NanoPlot --help
usage: NanoPlot [-h] [-v] [-t THREADS] [--verbose] [--store] [--raw] [--huge]
                [-o OUTDIR] [-p PREFIX] [--tsv_stats] [--maxlength N]
                [--minlength N] [--drop_outliers] [--downsample N]
                [--loglength] [--percentqual] [--alength] [--minqual N]
                [--runtime_until N] [--readtype {1D,2D,1D2}] [--barcoded]
                [--no_supplementary] [-c COLOR] [-cm COLORMAP]
                [-f {eps,jpeg,jpg,pdf,pgf,png,ps,raw,rgba,svg,svgz,tif,tiff}]
                [--plots [{kde,hex,dot,pauvre} [{kde,hex,dot,pauvre} ...]]]
                [--listcolors] [--listcolormaps] [--no-N50] [--N50]
                [--title TITLE] [--font_scale FONT_SCALE] [--dpi DPI]
                [--hide_stats]
                (--fastq file [file ...] | --fasta file [file ...] | --fastq_rich file [file ...] | --fastq_minimal file [file ...] | --summary file [file ...] | --bam file [file ...] | --ubam file [file ...] | --cram file [file ...] | --pickle pickle | --feather file [file ...])

CREATES VARIOUS PLOTS FOR LONG READ SEQUENCING DATA.
```

**This mean the environment works and we can use NanoPlot now.**

* We will use the [NanoPlot](https://github.com/wdecoster/NanoPlot) and some Rscripts to produce some plots of the basic stats of our data. To use Nanoplot we need first to request an interactive job and start working there (Remember: Never run a job in the login node "Do not get naked in the lobby"). 
*As some times we could lost the connection with the Orion server, it is preferable to use [TMUX](https://github.com/tmux/tmux/wiki). This terminal multiplexer software will allow keeping our session active in Orion even if we suddenly lost the internet connection:*

```console
[bio326-21-0@login GenomeAssembly2022]$ tmux new -s ONP
Welcome to the NMBU Orion compute cluster environment.

You are logged in to a machine that can be used to access your home directory,
edit your scripts, manage your files, and submit jobs to the cluster environment.
Do not run any jobs on this machine, as they might be automatically terminated.

IMPORTANT:
  - Orion introduction: https://orion.nmbu.no/
  - Orion can handle small-scale projects. Need more CPU hours? Please consider
    applying for national infrastructure resources: https://www.sigma2.no/
  - Please, PLEASE do compress your fastq, vcf and other non-compressed files
    using i.e. pigz.

For any Orion related enquiry: orion-support@nmbu.no
PS: We are on Teams: https://bit.ly/orion-teams

[bio326-21-0@login GenomeAssembly2022]$
```
* Now we need to ask the SLURM manager for a computer node to test the NANOPLOT and can start to work: 

```console 
[bio326-21-0@login GenomeAssembly2022]$ srun --cpus-per-task 4 --nodes 1 --mem=10G --time=02:00:00 --pty bash -i
srun: job 14301977 queued and waiting for resources
srun: job 14301977 has been allocated resources

Welcome to the NMBU Orion compute cluster environment.

You are logged in to a machine that can be used to access your home directory,
edit your scripts, manage your files, and submit jobs to the cluster environment.
Do not run any jobs on this machine, as they might be automatically terminated.

IMPORTANT:
  - Orion introduction: https://orion.nmbu.no/
  - Orion can handle small-scale projects. Need more CPU hours? Please consider
    applying for national infrastructure resources: https://www.sigma2.no/
  - Please, PLEASE do compress your fastq, vcf and other non-compressed files
    using i.e. pigz.

NEWS:
  - 2020-10-08: Orion has been re-built. We are still working out many details.
    Please email us if you miss anything, or notice any issues.

For any Orion related enquiry: orion-support@nmbu.no
PS: We are on Teams: https://bit.ly/orion-teams

[bio326-21-0@cn-12 GenomeAssembly2022]$
```
*Now that TMUX is active you can deattach the temrinal and try if your session is still active. To do this we need to press the keys ctrl+b d. This will bring us back to the "main" terminal...Use the command ```tmux a -t ONP``` to attach the ONP session again*

```console
[bio326-21-0@login GenomeAssembly2022]$ tmux a -t ONP
[bio326-21-0@cn-11 GenomeAssembly2022]$
```

As we moved into a new computer, we should activate Anaconda3 and the ONPTools environment before start working:

```console
[bio326-21-0@cn-11 GenomeAssembly2022]$ source /mnt/SCRATCH/bio326-21/GenomeAssembly/condaenvironments/activate.conda.sh
Activating Anaconda module for bio326-21-0
conda is running. Please type conda activate to load the basic conda functions...
[bio326-21-0@cn-11 GenomeAssembly2022]$ conda activate  /mnt/SCRATCH/bio326-21/GenomeAssembly/condaenvironments/ONPTools/
(/mnt/SCRATCH/bio326-21/GenomeAssembly/condaenvironments/ONPTools) [bio326-21-0@cn-11 GenomeAssembly2022]$
```

Let's run NanoPlot to generate some nice Plots. First let's run on the MiniONReads:

```console
(/mnt/SCRATCH/bio326-21/GenomeAssembly/condaenvironments/ONPTools) [bio326-21-0@cn-11 GenomeAssembly2022]$ cd MiniONReads/
(/mnt/SCRATCH/bio326-21/GenomeAssembly/condaenvironments/ONPTools) [bio326-21-0@cn-11 MiniONReads]$ ls
MiniON.fastq
(/mnt/SCRATCH/bio326-21/GenomeAssembly/condaenvironments/ONPTools) [bio326-21-0@cn-11 MiniONReads]$ NanoPlot -t 4 --fastq MiniON.fastq --N50 --loglength -p MiniON. -o MiniON.Nanoplot.out
/net/cn-1/mnt/SCRATCH/bio326-21/GenomeAssembly/condaenvironments/ONPTools/lib/python3.6/_collections_abc.py:702: MatplotlibDeprecationWarning:

The global colormaps dictionary is no longer considered public API.

/net/cn-1/mnt/SCRATCH/bio326-21/GenomeAssembly/condaenvironments/ONPTools/lib/python3.6/_collections_abc.py:720: MatplotlibDeprecationWarning:

The global colormaps dictionary is no longer considered public API.
(/mnt/SCRATCH/bio326-21/GenomeAssembly/condaenvironments/ONPTools) [bio326-21-0@cn-11 MiniONReads]$ ls
MiniON.fastq  MiniON.Nanoplot.out
```

The results are storage in the MiniON.Nanoplot.out directory. Take a look:

```console
(/mnt/SCRATCH/bio326-21/GenomeAssembly/condaenvironments/ONPTools) [bio326-21-0@cn-11 MiniONReads]$ cd MiniON.Nanoplot.out/
(/mnt/SCRATCH/bio326-21/GenomeAssembly/condaenvironments/ONPTools) [bio326-21-0@cn-11 MiniON.Nanoplot.out]$ ls
MiniON.Dynamic_Histogram_Read_length.html  MiniON.LengthvsQualityScatterPlot_loglength_dot.png  MiniON.NanoPlot-report.html                             MiniON.Yield_By_Length.png
MiniON.HistogramReadlength.png             MiniON.LengthvsQualityScatterPlot_loglength_kde.png  MiniON.NanoStats.txt
MiniON.LengthvsQualityScatterPlot_dot.png  MiniON.LogTransformed_HistogramReadlength.png        MiniON.Weighted_HistogramReadlength.png
MiniON.LengthvsQualityScatterPlot_kde.png  MiniON.NanoPlot_20220317_1748.log                    MiniON.Weighted_LogTransformed_HistogramReadlength.png
```

There are plenty of tables, pictures and reports. We can do a qick check of the NanoStats.txt report:

```console
(/mnt/SCRATCH/bio326-21/GenomeAssembly/condaenvironments/ONPTools) [bio326-21-0@cn-11 MiniON.Nanoplot.out]$ more MiniON.NanoStats.txt
General summary:
Mean read length:                4,266.4
Mean read quality:                  10.2
Median read length:              2,275.5
Median read quality:                10.5
Number of reads:                24,000.0
Read length N50:                 8,318.0
STDEV read length:               5,682.0
Total bases:               102,392,823.0
Number, percentage and megabases of reads above quality cutoffs
>Q5:    23999 (100.0%) 102.4Mb
>Q7:    22812 (95.0%) 98.1Mb
>Q10:   14172 (59.0%) 62.6Mb
>Q12:   3881 (16.2%) 16.7Mb
>Q15:   0 (0.0%) 0.0Mb
Top 5 highest mean basecall quality scores and their read lengths
1:      14.8 (962)
2:      14.8 (772)
3:      14.5 (1027)
4:      14.4 (1483)
5:      14.4 (894)
Top 5 longest reads and their mean basecall quality score
1:      117332 (11.8)
2:      82149 (9.4)
3:      68986 (7.5)
4:      66931 (8.2)
5:      63998 (8.6)

```

Let's do the same for the PromethiON results:

```console
(/mnt/SCRATCH/bio326-21/GenomeAssembly/condaenvironments/ONPTools) [bio326-21-0@cn-11 MiniON.Nanoplot.out]$ cd /mnt/SCRATCH/bio326-21-0/GenomeAssembly2022
(/mnt/SCRATCH/bio326-21/GenomeAssembly/condaenvironments/ONPTools) [bio326-21-0@cn-11 GenomeAssembly2022]$ cd PromethiONReads/
(/mnt/SCRATCH/bio326-21/GenomeAssembly/condaenvironments/ONPTools) [bio326-21-0@cn-11 PromethiONReads]$ NanoPlot -t 4 --fastq PromethiON.fastq --N50 --loglength -p PromethiON. -o PromethiON.Nanoplot.out
/net/cn-1/mnt/SCRATCH/bio326-21/GenomeAssembly/condaenvironments/ONPTools/lib/python3.6/_collections_abc.py:702: MatplotlibDeprecationWarning:

The global colormaps dictionary is no longer considered public API.

/net/cn-1/mnt/SCRATCH/bio326-21/GenomeAssembly/condaenvironments/ONPTools/lib/python3.6/_collections_abc.py:720: MatplotlibDeprecationWarning:

The global colormaps dictionary is no longer considered public API.
(/mnt/SCRATCH/bio326-21/GenomeAssembly/condaenvironments/ONPTools) [bio326-21-0@cn-11 PromethiON.Nanoplot.out]$ cd PromethiON.Nanoplot.out/
(/mnt/SCRATCH/bio326-21/GenomeAssembly/condaenvironments/ONPTools) [bio326-21-0@cn-11 PromethiON.Nanoplot.out]$ more PromethiON.NanoStats.txt
General summary:
Mean read length:                 6,935.2
Mean read quality:                   12.6
Median read length:               2,029.0
Median read quality:                 12.7
Number of reads:                165,305.0
Read length N50:                 23,747.0
STDEV read length:               12,384.9
Total bases:              1,146,423,404.0
Number, percentage and megabases of reads above quality cutoffs
>Q5:    165305 (100.0%) 1146.4Mb
>Q7:    165305 (100.0%) 1146.4Mb
>Q10:   135154 (81.8%) 934.6Mb
>Q12:   99363 (60.1%) 715.0Mb
>Q15:   29523 (17.9%) 208.5Mb
Top 5 highest mean basecall quality scores and their read lengths
1:      22.9 (629)
2:      22.6 (282)
3:      22.1 (219)
4:      21.5 (262)
5:      21.5 (1414)
Top 5 longest reads and their mean basecall quality score
1:      261876 (7.6)
2:      240794 (8.6)
3:      196851 (8.0)
4:      189140 (7.9)
5:      184049 (9.3)

```

NanoPlot also proudces a .html report with these data. To vizualise this it is necesary to copy the html to our computer and open it then in a browser.
We can do this by scp or rsync commands **IN YOUR COMPUTER NOT ORION**:

```console
(base) avera@L003772:auve$ rsync -aP bio326-21-0@login.orion.nmbu.no:/mnt/SCRATCH/bio326-21-0/GenomeAssembly2022/PromethiONReads/PromethiON.Nanoplot.out/*.html .
```

What can we conclude on these results???

Now we can finish the interactive job and the tmux session:

```console
(/mnt/SCRATCH/bio326-21/GenomeAssembly/condaenvironments/ONPTools) [bio326-21-0@cn-11 PromethiON.Nanoplot.out]$ exit
(/mnt/SCRATCH/bio326-21/GenomeAssembly/condaenvironments/ONPTools) [bio326-21-0@login Nanoplot.out]$ exit
[exited]
```

## Filtering lowquality reads and short reads by [FiltLong](https://github.com/rrwick/Filtlong)

From the FiltLong github:
*"Filtlong is a tool for filtering long reads by quality. It can take a set of long reads and produce a smaller, better subset. It uses both read length (longer is better) and read identity (higher is better) when choosing which reads pass the filter."*

We will use this software to remove those "bad" reads and clean the samples.

The easiest way is by submmiting the following script as SLRUM batch job:

```bash
#!/bin/bash
#########################################################################
#       SLURM script to run FiltLong and NanoPlot analysis on ONP reads
#               usage: sbatch FiltLong.SLURM.sh <InputString>
#
#       Author Arutro Vera
#       March 2022
#########################################################################

###############SLURM SCRIPT###################################

## Job name:
#SBATCH --job-name=FiltlongAndNanoPlot
#
## Wall time limit:
#SBATCH --time=01:00:00
#
## Other parameters:
#SBATCH --cpus-per-task 10
#SBATCH --mem=20G
#SBATCH --out slurm-FiltlongNanoplot-%A.out
#SBATCH --partition=smallmem
###
###########################################################

###Basic usage help for this script#######

print_usage() {
        echo "Usage: sbatch $0 <InputString>"
        echo "eg: sbatch $0 MiniON"
}

if [ $# -le 0 ]
        then
                print_usage
                exit 1
        fi


###############Main SCRIPT###################################

###Variables###
input=$1


## Set up job environment:

module --quiet purge  # Reset the modules to the system default
module load Miniconda3

##Activate conda environments

export PS1=\$

source /mnt/SCRATCH/bio326-21/GenomeAssembly/condaenvironments/activate.conda.sh
conda activate /mnt/SCRATCH/bio326-21/GenomeAssembly/condaenvironments/ONPTools/


####Do some work:########

## For debuggin
echo "Hello" $USER
echo "my submit directory is:"
echo $SLURM_SUBMIT_DIR
echo "this is the job:"
echo $SLURM_JOB_ID
echo "I am running on:"
echo $SLURM_NODELIST
echo "I am running with:"
echo $SLURM_CPUS_ON_NODE "cpus"
echo "I'm working with this CONDAENV"
echo $CONDA_PREFIX
echo "Today is:"
date

## Copying data to local node for faster computation

cd $TMPDIR

#Check if $USER exists in $TMPDIR

if [[ -d $USER ]]
        then
                echo "$USER exists on $TMPDIR"
        else
                mkdir $USER
fi


echo "copying files to" $TMPDIR/$USER/tmpDir_of.$SLURM_JOB_ID.$input

cd $USER
mkdir tmpDir_of.$SLURM_JOB_ID.$input
cd tmpDir_of.$SLURM_JOB_ID.$input

#Copy data to the $TMPDIR

echo "copying Reads to" $TMPDIR/$USER/tmpDir_of.$SLURM_JOB_ID.$input

time rsync -aP $SLURM_SUBMIT_DIR/$input\Reads/*fastq .

echo "These are my files:"
ls -1


### Filtlong

echo "Performing Filtlong on file $input ..."
date +%d\ %b\ %T

time filtlong \
        --min_length 1000 \
        --keep_percent 90 \
        $input.fastq | pigz -p $SLURM_CPUS_ON_NODE > $input.filtlong.fq.gz

###NanoPlot

echo "Performing NanoPlot on filtlong reads"
date +%d\ %b\ %T

time NanoPlot \
        -t $SLURM_CPUS_ON_NODE \
        --fastq $input.filtlong.fq.gz \
        --N50 \
        --loglength \
        -p $input.filtlong. \
        -o $input.NanoplotFiltlong.dir

###########Moving results to anywhere the main script was submitted############

echo "moving results to" $SLURM_SUBMIT_DIR/

cd $TMPDIR/$USER/tmpDir_of.$SLURM_JOB_ID.$input

time rsync -aP $input.filtlong.fq.gz $SLURM_SUBMIT_DIR/$input\Reads/
time rsync -aP $input.NanoplotFiltlong.dir $SLURM_SUBMIT_DIR/$input\Reads

echo "Final fastq and Nanoplot results are in:" $SLURM_SUBMIT_DIR/$input\Reads

####removing tmp dir. Remember to do this for not filling the HDD in the node!!!!###

cd $TMPDIR/$USER/
rm -r tmpDir_of.$SLURM_JOB_ID.$input

echo "I've done at"
date
```

A copy of this script can be found in ```/mnt/SCRATCH/bio326-21/GenomeAssembly/BIO326-2022/scripts``` Copy to your ```GenomeAssembly2022``` directory as follow:

```console
[bio326-21-0@login GenomeAssembly2022]$ cp /mnt/SCRATCH/bio326-21/GenomeAssembly/BIO326-2022/scripts/FiltLong.SLURM.sh .
[bio326-21-0@login GenomeAssembly2022]$ ls
FiltLong.SLURM.sh  MiniONReads  PromethiONReads
```
This script has a usage lets check the options:

```console 
[bio326-21-0@login GenomeAssembly2022]$ ./FiltLong.SLURM.sh
Usage: sbatch ./FiltLong.SLURM.sh <InputString>
eg: sbatch ./FiltLong.SLURM.sh MiniON
```
And follow the example:

```console
[bio326-21-0@login GenomeAssembly2022]$ sbatch FiltLong.SLURM.sh MiniON
Submitted batch job 14302187
```

We can monitorate the progress of our job by using the ```squeue``` command:

```console
[bio326-21-0@login GenomeAssembly2022]$ squeue -u $USER
             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
          14302187  smallmem Filtlong bio326-2  R       0:10      1 cn-8
```
The same way we can run the script for the Promethion:

```console
[bio326-21-0@login GenomeAssembly2022]$ sbatch FiltLong.SLURM.sh PromethiON
Submitted batch job 14302190
[bio326-21-0@login GenomeAssembly2022]$ squeue -u $USER
             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
          14302190  smallmem Filtlong bio326-2  R       0:01      1 cn-8
```

SBATCH will produce an .out file with all the messages from the STDOUT and errors, let's take a look:

```console
[bio326-21-0@login GenomeAssembly2022]$ ls -1
FiltLong.SLURM.sh
MiniONReads
PromethiONReads
slurm-FiltlongNanoplot-14302187.out
slurm-FiltlongNanoplot-14302190.out
[bio326-21-0@login GenomeAssembly2022]$ head -20 slurm-FiltlongNanoplot-14302187.out
Activating Anaconda module for bio326-21-0
conda is running. Please type conda activate to load the basic conda functions...
Hello bio326-21-0
my submit directory is:
/net/cn-1/mnt/SCRATCH/bio326-21-0/GenomeAssembly2022
this is the job:
14302187
I am running on:
cn-8
I am running with:
10 cpus
I'm working with this CONDAENV
/mnt/SCRATCH/bio326-21/GenomeAssembly/condaenvironments/ONPTools
Today is:
Thu Mar 17 21:19:03 CET 2022
copying files to /home/work/bio326-21-0/tmpDir_of.14302187.MiniON
copying Reads to /home/work/bio326-21-0/tmpDir_of.14302187.MiniON
sending incremental file list
MiniON.fastq
    208,625,216 100%  104.53MB/s    0:00:01 (xfr#1, to-chk=0/1)
  ```
 
 We can display also the last part of the file:
 
 ```console
  [bio326-21-0@login GenomeAssembly2022]$ tail slurm-FiltlongNanoplot-14302190.out
         28,880 100%  273.82kB/s    0:00:00 (xfr#12, to-chk=1/14)
PromethiON.NanoplotFiltlong.dir/PromethiON.filtlong.Yield_By_Length.png
         22,684 100%  215.07kB/s    0:00:00 (xfr#13, to-chk=0/14)

real    0m0.162s
user    0m0.042s
sys     0m0.021s
Final fastq and Nanoplot results are in: /net/cn-1/mnt/SCRATCH/bio326-21-0/GenomeAssembly2022/PromethiONReads
I've done at
Thu Mar 17 21:24:51 CET 2022
```
It looks like our script worked okay. Accordingly to the slurm-out file the results are in ```/net/cn-1/mnt/SCRATCH/bio326-21-0/GenomeAssembly2022/PromethiONReads```. Let's go there and see what we got:

```console
[bio326-21-0@login GenomeAssembly2022]$ cd /net/cn-1/mnt/SCRATCH/bio326-21-0/GenomeAssembly2022/PromethiONReads
[bio326-21-0@login PromethiONReads]$ ls
PromethiON.fastq  PromethiON.filtlong.fq.gz  PromethiON.NanoplotFiltlong.dir  PromethiON.Nanoplot.out
[bio326-21-0@login PromethiONReads]$ cd PromethiON.NanoplotFiltlong.dir/
[bio326-21-0@login PromethiON.NanoplotFiltlong.dir]$ ls
PromethiON.filtlong.Dynamic_Histogram_Read_length.html            PromethiON.filtlong.NanoPlot_20220317_2122.log
PromethiON.filtlong.HistogramReadlength.png                       PromethiON.filtlong.NanoPlot-report.html
PromethiON.filtlong.LengthvsQualityScatterPlot_dot.png            PromethiON.filtlong.NanoStats.txt
PromethiON.filtlong.LengthvsQualityScatterPlot_kde.png            PromethiON.filtlong.Weighted_HistogramReadlength.png
PromethiON.filtlong.LengthvsQualityScatterPlot_loglength_dot.png  PromethiON.filtlong.Weighted_LogTransformed_HistogramReadlength.png
PromethiON.filtlong.LengthvsQualityScatterPlot_loglength_kde.png  PromethiON.filtlong.Yield_By_Length.png
PromethiON.filtlong.LogTransformed_HistogramReadlength.png
```
It seems everything works, let's display the NanoStats:

```console
[bio326-21-0@login PromethiON.NanoplotFiltlong.dir]$ more PromethiON.filtlong.NanoStats.txt
eneral summary:
Mean read length:                11,276.5
Mean read quality:                   13.3
Median read length:               5,085.0
Median read quality:                 13.5
Number of reads:                 91,499.0
Read length N50:                 24,931.0
STDEV read length:               14,439.0
Total bases:              1,031,784,865.0
Number, percentage and megabases of reads above quality cutoffs
>Q5:    91499 (100.0%) 1031.8Mb
>Q7:    91499 (100.0%) 1031.8Mb
>Q10:   84389 (92.2%) 905.2Mb
>Q12:   66622 (72.8%) 696.9Mb
>Q15:   21354 (23.3%) 203.9Mb
Top 5 highest mean basecall quality scores and their read lengths
1:      21.5 (1414)
2:      20.8 (1245)
3:      20.6 (1252)
4:      20.5 (1387)
5:      20.3 (2425)
Top 5 longest reads and their mean basecall quality score
1:      240794 (8.6)
2:      196851 (8.
```
We can compare this with the "raw" PromethiON NanoStats:

```console
[bio326-21-0@login PromethiON.NanoplotFiltlong.dir]$ more ../PromethiON.Nanoplot.out/PromethiON.NanoStats.txt
General summary:
Mean read length:                 6,935.2
Mean read quality:                   12.6
Median read length:               2,029.0
Median read quality:                 12.7
Number of reads:                165,305.0
Read length N50:                 23,747.0
STDEV read length:               12,384.9
Total bases:              1,146,423,404.0
Number, percentage and megabases of reads above quality cutoffs
>Q5:    165305 (100.0%) 1146.4Mb
>Q7:    165305 (100.0%) 1146.4Mb
>Q10:   135154 (81.8%) 934.6Mb
>Q12:   99363 (60.1%) 715.0Mb
>Q15:   29523 (17.9%) 208.5Mb
Top 5 highest mean basecall quality scores and their read lengths
1:      22.9 (629)
2:      22.6 (282)
3:      22.1 (219)
4:      21.5 (262)
5:      21.5 (1414)
Top 5 longest reads and their mean basecall quality score
1:      261876 (7.6)
2:      240794 (8.6)
3:      196851 (8.0)
4:      189140 (7.9)
5:      184049 (9.3)
```
It seems there was a little improvment after FiltLong selection. We could keep opening all the NanoStats files and compare manualy, this is easy when you have 2 or 3 files but what happen if you get 100? Manual inspection will take forever. So as bioinformaticians we can create our own scripts to parse the information from NanoPlot and display all this info in some kind of plots.

* Go to the GenomeAssembly2022 directory and let's gather all the NanoStatsFiles into a single directory (e.g ```NanoStats.dir```). Again a ```for loop``` will be super useful here:

```console
bio326-21-0@login PromethiON.NanoplotFiltlong.dir]$ cd $SCRATCH/GenomeAssembly2022
[bio326-21-0@login NanoStats.dir]$ mkdir NanoStats.dir
[bio326-21-0@login GenomeAssembly2022]$ for i in *Reads; do BN=$(basename $i Reads); cp $i/$BN.Nanoplot.out/*NanoStats.txt NanoStats.dir/; cp $i/$BN.NanoplotFiltlong.dir/*NanoStats.txt NanoStats.dir/;done
```
* Then got to the new directory and check if our ```for loop``` worked:

```console
[bio326-21-0@login GenomeAssembly2022]$ cd NanoStats.dir/
[bio326-21-0@login NanoStats.dir]$ ls
MiniON.filtlong.NanoStats.txt  MiniON.NanoStats.txt  PromethiON.filtlong.NanoStats.txt  PromethiON.NanoStats.txt
```

We will use some scripts to perform some task so let's ask for a computing node and not work in the loggin node:

```console
[bio326-21-0@login NanoStats.dir]$ srun --cpus-per-task 4 --nodes 1 --mem=10G --time=02:00:00 --pty bash -i
srun: job 14302254 queued and waiting for resources
srun: job 14302254 has been allocated resources

Welcome to the NMBU Orion compute cluster environment.

You are logged in to a machine that can be used to access your home directory,
edit your scripts, manage your files, and submit jobs to the cluster environment.
Do not run any jobs on this machine, as they might be automatically terminated.

IMPORTANT:
  - Orion introduction: https://orion.nmbu.no/
  - Orion can handle small-scale projects. Need more CPU hours? Please consider
    applying for national infrastructure resources: https://www.sigma2.no/
  - Please, PLEASE do compress your fastq, vcf and other non-compressed files
    using i.e. pigz.

NEWS:
  - 2020-10-08: Orion has been re-built. We are still working out many details.
    Please email us if you miss anything, or notice any issues.

For any Orion related enquiry: orion-support@nmbu.no
PS: We are on Teams: https://bit.ly/orion-teams

[bio326-21-0@cn-11 NanoStats.dir]$
```
* The script ```/mnt/SCRATCH/bio326-21/GenomeAssembly/BIO326-2022/scripts/PNanostats.2.pl``` will help us to gather all the tables into a single file and extract the No. of reads, the mean lenght, N50 and the longest read in each file. Let's use it:

```console
(/mnt/SCRATCH/bio326-21/GenomeAssembly/condaenvironments/r-env) [bio326-21-0@cn-11 NanoStats.dir]$ perl /mnt/SCRATCH/bio326-21/GenomeAssembly/BIO326-2022/scripts/PNanostats.2.pl
Reading File MiniON.NanoStats.txt
Reading File PromethiON.NanoStats.txt
Reading File MiniON.filtlong.NanoStats.txt
Reading File PromethiON.filtlong.NanoStats.txt
The result table is NanoStats.total.tsv

I've done.
Thank you!
```
* After cheking we got the ```NanoStats.total.tsv``` file in our directory we can exit the computing node:

```console
[bio326-21-0@cn-11 NanoStats.dir]$ ls -lrth
total 20K
-rw-rw-r-- 1 bio326-21-0 bio326-21-0 819 Mar 17 22:03 MiniON.NanoStats.txt
-rw-rw-r-- 1 bio326-21-0 bio326-21-0 812 Mar 17 22:03 MiniON.filtlong.NanoStats.txt
-rw-rw-r-- 1 bio326-21-0 bio326-21-0 846 Mar 17 22:03 PromethiON.NanoStats.txt
-rw-rw-r-- 1 bio326-21-0 bio326-21-0 847 Mar 17 22:03 PromethiON.filtlong.NanoStats.txt
-rw-rw-r-- 1 bio326-21-0 bio326-21-0 186 Mar 18 10:58 NanoStats.total.tsv
[bio326-21-0@cn-11 NanoStats.dir]$ more NanoStats.total.tsv
MiniON  4,266.4 24,000.0        8,318.0 117332
PromethiON      6,935.2 165,305.0       23,747.0        261876
MiniON.filtlong 6,836.8 13,479.0        9,438.0 117332
PromethiON.filtlong     11,276.5        91,499.0        24,931.0        240794
[bio326-21-0@cn-11 NanoStats.dir]$ exit
exit
[bio326-21-0@login NanoStats.dir]$
```

* Finally we can use R (RStudio is a nice option) to produce the plots. To do this we can open RStudio in the jupyterHub: https://orion.nmbu.no/jupyter/ Select the **RStudio-4.0.5** ![JP](https://github.com/TheMEMOLab/Bio326-NMBU/blob/main/images/jpyter.png).


* We can use the following script to produce nice barplots of the NanoStats

```R

###Addign a path with some libraries used in BIO326
mypath <- "/mnt/SCRATCH/bio326-21/GenomeAssembly/R"
.libPaths(mypath)

##Loading libraries
library(tidyverse)
library(patchwork)

##Go to the GenomeAssembly2022 dir

setwd("/mnt/SCRATCH/bio326-21-0/GenomeAssembly2022/NanoStats.dir")

#Reading the NanoStats table

Nanostat <- read_tsv("NanoStats.total.tsv",
                     col_names = c("Sample",
                                   "Mean.Read.Len", 
                                   "Num.Reads",
                                   "N50",
                                   "Quality",
                                   "Longest"))

###USing ggplot to create barplots of the stats

BPNReads <- ggplot(data=Nanostat,
                   aes(x=reorder(Sample,Num.Reads),
                       y=Num.Reads)) +
  geom_bar(stat='identity',fill="steelblue")+
  xlab("Samples")+
  ylab("Number of reads")+
  coord_flip()+
  theme_classic()


BPN50 <- ggplot(data=Nanostat,
                aes(x=reorder(Sample,N50),
                    y=N50)) +
  geom_bar(stat='identity',fill="darkgreen")+
  coord_flip()+
  xlab("Samples")+
  ylab("N50")+
  theme_classic()

BPLonges <- ggplot(data=Nanostat,
                   aes(x=reorder(Sample,Longest),
                       y=Longest)) +
  geom_bar(stat='identity',fill="darkred")+
  coord_flip()+
  xlab("Samples")+
  ylab("longestRead_Length_bp")+
  theme_classic()


BPMeanLen <- ggplot(data=Nanostat,
                    aes(x=reorder(Sample,Mean.Read.Len),
                        y=Mean.Read.Len)) +
  geom_bar(stat='identity',fill="orange")+
  coord_flip()+
  xlab("Samples")+
  ylab("Mean_Read_Length_bp")+
  theme_classic()

Quality <- ggplot(data=Nanostat,
                  aes(x=reorder(Sample,Quality),
                      y=Quality)) +
  geom_bar(stat='identity',fill="pink")+
  coord_flip()+
  xlab("Samples")+
  ylab("Mean read Quality")+
  theme_classic()

##Using patchwork to merge all the ggplot objects into a single object

BarplotsAll <- (BPNReads + BPN50+Quality)/(BPMeanLen + BPLonges)

##Saving the BarplotsAll object into a PDF

ggsave(BarplotsAll,
       file="Barplots.NanoStats.pdf",
       width = 15,
       height = 18)

```


* This script produces a pdf with the plots. To open it, we need to transfer the file to our computer by rsync: 

```console
(base) avera@L003772:~$ rsync -aP bio326-21-0@login.orion.nmbu.no:/mnt/SCRATCH/bio326-21-0/GenomeAssembly2022/NanoStats.dir/*.pdf .
bio326-21-0@login.orion.nmbu.no's password:
receiving incremental file list
Barplots.NanoStats.pdf
          5,288 100%    5.04MB/s    0:00:00 (xfr#1, to-chk=0/1)
```

The PDF shows something like this ![barplot](https://github.com/TheMEMOLab/Bio326-NMBU/blob/main/images/barplot.png)

**Do you think FiltLong improve the quality of the Nanopore sequences????**

## Genome assembly using [flye](https://github.com/fenderglass/Flye)

Now that we have filtered the reads using Filtlong the next step is assembly the reads into genomic contigs.

* We will use Flye assembler to merge the reads into a genome assembly.

* The following script uses the reads filtered by filtlong and then call Flye to produce a genomic assembly:

```bash

#!/bin/bash
#########################################################################
#       SLURM script to run Flye ONP reads
#               usage: sbatch Flye.SLURM.sh <InputString>
#
#       Author Arutro Vera
#       March 2022
#########################################################################

###############SLURM SCRIPT###################################

## Job name:
#SBATCH --job-name=FlyeAssembly
#
## Wall time limit:
#SBATCH --time=24:00:00
#
## Other parameters:
#SBATCH --cpus-per-task 12
#SBATCH --mem=50G
#SBATCH --out slurm-FlyeAssembly-%A.out
#SBATCH --partition=smallmem
###
###########################################################

###Basic usage help for this script#######

print_usage() {
        echo "Usage: sbatch $0 <InputString>"
        echo "eg: sbatch $0 MiniON"
}

if [ $# -le 0 ]
        then
                print_usage
                exit 1
        fi


###############Main SCRIPT###################################

###Variables###
input=$1


## Set up job environment:

module --quiet purge  # Reset the modules to the system default
module load Miniconda3

##Activate conda environments

export PS1=\$

source /mnt/SCRATCH/bio326-21/GenomeAssembly/condaenvironments/activate.conda.sh
conda activate /net/cn-1/mnt/SCRATCH/bio326-21/GenomeAssembly/condaenvironments/ONPTools/ONPAssembly


####Do some work:########

## For debuggin
echo "Hello" $USER
echo "my submit directory is:"
echo $SLURM_SUBMIT_DIR
echo "this is the job:"
echo $SLURM_JOB_ID
echo "I am running on:"
echo $SLURM_NODELIST
echo "I am running with:"
echo $SLURM_CPUS_ON_NODE "cpus"
echo "I'm working with this CONDAENV"
echo $CONDA_PREFIX
echo "Today is:"
date

## Copying data to local node for faster computation

cd $TMPDIR

#Check if $USER exists in $TMPDIR

if [[ -d $USER ]]
        then
                echo "$USER exists on $TMPDIR"
        else
                mkdir $USER
fi


echo "copying files to" $TMPDIR/$USER/tmpDir_of.$SLURM_JOB_ID.$input

cd $USER
mkdir tmpDir_of.$SLURM_JOB_ID.$input
cd tmpDir_of.$SLURM_JOB_ID.$input

#Copy data to the $TMPDIR

echo "copying Filtlong reads to" $TMPDIR/$USER/tmpDir_of.$SLURM_JOB_ID.$input

time rsync -aP $SLURM_SUBMIT_DIR/$input\Reads/$input.filtlong.fq.gz .

echo "These are my files:"
ls -1

##Assembly using flye

echo "Assembly bacterial genome $input using Fly"
date +%d\ %b\ %T

time flye \
        --nano-raw $input.filtlong.fq.gz \
        --threads $SLURM_CPUS_ON_NODE \
        -o $input.flyeAssembly.dir

echo "Doing some stats..."

### Stats

conda activate /net/cn-1/mnt/SCRATCH/bio326-21/GenomeAssembly/condaenvironments/ONPTools/

cd $input.flyeAssembly.dir

echo "stats"
assembly-stats assembly.fasta
Genome_Assembly_lecture/Scripts/assembly-stats/assembly-stats assembly.fasta > $input.stats.metaFly.contigs.txt

###########Moving results to anywhere the main script was submitted############

echo "moving results to" $SLURM_SUBMIT_DIR/

cd $TMPDIR/$USER/tmpDir_of.$SLURM_JOB_ID.$input

time rsync -aP $input.flyeAssembly.dir $SLURM_SUBMIT_DIR/

echo "Final fastq and Nanoplot results are in:" $SLURM_SUBMIT_DIR/$input.flyeAssembly.dir

####removing tmp dir. Remember to do this for not filling the HDD in the node!!!!###

cd $TMPDIR/$USER/
rm -r tmpDir_of.$SLURM_JOB_ID.$input

echo "I've done at"
date

```
* You can copy the script from ```/mnt/SCRATCH/bio326-21/GenomeAssembly/BIO326-2022/scripts```

```console 
[bio326-21-0@login GenomeAssembly2022]$ cp /mnt/SCRATCH/bio326-21/GenomeAssembly/BIO326-2022/scripts/flye.SLURM.sh .
```

* And run like this:

```console
[bio326-21-0@login GenomeAssembly2022]$ sbatch flye.SLURM.sh MiniON
```
** Due to the lack of time, in class we only work with the MiniON files. As homework you will assembly the PromethiON reads.**

**The assembly will take some time...let's check some concepts:

![Assembly](https://github.com/TheMEMOLab/Bio326-NMBU/blob/main/images/assembly.png)
![Stats](https://github.com/TheMEMOLab/Bio326-NMBU/blob/main/images/stats.png)

* The script will produce the ```MiniON.flyeAssembly.dir``` folder:

```console
(/mnt/users/auve/mycondaenvs/ONPTools) [bio326-21-0@login GenomeAssembly2022]$ ls -1
FiltLong.SLURM.sh
flye.SLURM.sh
MiniON.flyeAssembly.dir
```
* If everything worked correctly you will obtain somethig like this:

```console
[bio326-21-0@login GenomeAssembly2022]$ tree MiniON.flyeAssembly.dir/
MiniON.flyeAssembly.dir/
├── 00-assembly
│   ├── draft_assembly.fasta
│   └── draft_assembly.fasta.fai
├── 10-consensus
│   ├── consensus.fasta
│   ├── minimap.bam.bai
│   └── minimap.stderr
├── 20-repeat
│   ├── graph_after_rr.gv
│   ├── graph_before_rr.fasta
│   ├── graph_before_rr.gv
│   ├── read_alignment_dump
│   ├── repeat_graph_dump
│   └── repeat_graph_edges.fasta
├── 30-contigger
│   ├── contigs.fasta
│   ├── contigs.fasta.fai
│   ├── contigs_stats.txt
│   ├── graph_final.fasta
│   ├── graph_final.gfa
│   ├── graph_final.gv
│   └── scaffolds_links.txt
├── 40-polishing
│   ├── contigs_stats.txt
│   ├── edges_aln.bam.bai
│   ├── filtered_contigs.fasta
│   ├── filtered_contigs.fasta.fai
│   ├── filtered_stats.txt
│   ├── minimap_1.bam.bai
│   ├── minimap.stderr
│   └── polished_edges.gfa
├── assembly.fasta
├── assembly_graph.gfa
├── assembly_graph.gv
├── assembly_info.txt
├── flye.log
├── MiniON.stats.flye.contigs.txt
└── params.json

5 directories, 33 files
```

**The file assembly.fasta contain the final contigs and the file  MiniON.stats.flye.contigs.txt has the basic stats. Let's take a look:**

```console
[bio326-21-0@login GenomeAssembly2022]$ more MiniON.flyeAssembly.dir/MiniON.stats.flye.contigs.txt
stats for assembly.fasta
sum = 3415822, n = 7, ave = 487974.57, largest = 3305136
N50 = 3305136, n = 1
N60 = 3305136, n = 1
N70 = 3305136, n = 1
N80 = 3305136, n = 1
N90 = 3305136, n = 1
N100 = 4184, n = 7
N_count = 0
Gaps = 0
```


The total of bases in the assembly sum ~3.46 Mb, this is the final length of the assembly. The N50 (statistic that defines the assembly quality in terms of contiguity. It can be defined as: given a set of contigs, the N50 is the sequence length of the shortest contig at 50% of the total genome length...), is ~ 3.3 Mb in a single contig.

After this, we can assess the completeness of our genome. The most used strategy is to look for the presence of a set of single-copy orthologs genes commonly present in all bacteria (universal genes) and score the number of occurrences in our genome.

### Evaluate genome completeness by BUSCO

[BUSCO](https://busco.ezlab.org/) (Benchmarking Universal Single-Copy Orthologs) is a tool that attempts to provide a quantitative assessment of the completeness in terms of expected gene content of a genome assembly, transcriptome, or annotated gene set. The results are simplified into categories of Complete and single-copy, Complete and duplicated, Fragmented, or Missing BUSCOs.

This software looks for a certain number of orthologous genes (BUSCOs) on a database and compares the total of these ortholog genes present in the genome we would like to evaluate. Then, it estimates the completeness based on the presence, duplication, fragmentation, or absence of these BUSCOS. For example (raw example), if the BUSCO database has 10 genes and the software only finds 9 of them in the query genome it scores completeness of the genome at 90 %.

BUSCO has developed different databases with common universal orthologs clusters for several organisms:
![buscoimg](https://github.com/avera1988/Genome_Assembly_lecture/blob/master/images/busco.png)

Busco will predict genes in the assembly (by prodigal) and then look for the USCOs of a certain taxonomical lineage using hmmer. It automatically identifies the closest taxonomical lineage and then download the BUSCOs database, however you can indicate and narrow the BUSCOs sarch to a prokaryote or eukaryote database by using the **--auto-lineage-prok** flag. 

- BUSCO is installed in Orion by a singularity container. We can use the following command to look into the busco options:
